nomad/secure_variables_endpoint.go

```diff
@@ -367,3 +367,56 @@ func (sv *SecureVariables) decrypt(v *structs.SecureVariable) error {
 	v.EncryptedData = nil
 	return nil
 }
+
+// handleMixedAuthEndpoint is a helper to handle auth on RPC endpoints that can
+// either be called by external clients or by workload identity
+func (sv *SecureVariables) handleMixedAuthEndpoint(args structs.QueryOptions, cap, pathOrPrefix string) error {
+
+	// Perform the initial token resolution.
+	aclObj, err := sv.srv.ResolveToken(args.AuthToken)
+	if err == nil {
+		// Perform our ACL validation. If the object is nil, this means ACLs
+		// are not enabled, otherwise trigger the allowed namespace function.
+		if aclObj != nil {
+			if !aclObj.AllowNsOp(args.RequestNamespace(), cap) {
+				return structs.ErrPermissionDenied
+			}
+		}
+		return nil
+	}
+
+	// Attempt to verify the token as a JWT with a workload
+	// identity claim
+	claim, err := sv.srv.ResolveClaim(args.AuthToken)
+	if err != nil {
+		return structs.ErrPermissionDenied
+	}
+
+	store := sv.srv.fsm.State()
+	alloc, err := store.AllocByID(nil, claim.AllocationID)
+	if err != nil || alloc == nil || alloc.Job == nil {
+		return structs.ErrPermissionDenied
+	}
+
+	// the claims for terminal allocs are always treated as expired
+	if alloc.TerminalStatus() {
+		return structs.ErrPermissionDenied
+	}
+
+	if alloc.Job.Namespace != args.RequestNamespace() {
+		return structs.ErrPermissionDenied
+	}
+
+	parts := strings.Split(pathOrPrefix, "/")
+	expect := []string{"jobs", alloc.Job.ID, alloc.TaskGroup, claim.TaskName}
+	if len(parts) > len(expect) {
+		return structs.ErrPermissionDenied
+	}
+
+	for idx, part := range parts {
+		if part != expect[idx] {
+			return structs.ErrPermissionDenied
+		}
+	}
```

Comment:
Some comments around the purpose of pathOrPrefix either here and/or in the func definition would be helpful. It looks like the purpose is to lock token access down to just secure variable paths for this task?

Also I think this logic is equivalent to `if pathOrPrefix != strings.Join(expect) { return structs.ErrPermissionDenied }` although I think you manage to dodge making any memory allocations.

Reply:
Resolved in 23af20e29b2398437d86c45487d7bccf7f639cb3